Mathlib/Computability/NFA.lean

```diff
@@ -124,6 +144,52 @@ theorem mem_evalFrom_iff_exists {s : σ} {S : Set σ} {x : List α} :
   rw [evalFrom_eq_biUnion_singleton]
   simp
 
+variable (M) in
+/-- `M.acceptsFrom S` is the language of `x` such that there is an accept state
+in `M.evalFrom S x`. -/
+def acceptsFrom (S : Set σ) : Language α := {x | ∃ s ∈ M.accept, s ∈ M.evalFrom S x}
+
+theorem mem_acceptsFrom {S : Set σ} {x : List α} :
+    x ∈ M.acceptsFrom S ↔ ∃ s ∈ M.accept, s ∈ M.evalFrom S x := by
+  rfl
+
+@[simp]
+theorem mem_acceptsFrom_nil {S : Set σ} : [] ∈ M.acceptsFrom S ↔ ∃ s ∈ S, s ∈ M.accept := by
+  simp [mem_acceptsFrom]; tauto
+
+@[simp]
+theorem mem_acceptsFrom_cons {S : Set σ} {a : α} {x : List α} :
+    a :: x ∈ M.acceptsFrom S ↔ x ∈ M.acceptsFrom (M.stepSet S a) := by
+  simp [mem_acceptsFrom]
+
+@[simp]
+theorem mem_acceptsFrom_append {S : Set σ} {x y : List α} :
+    x ++ y ∈ M.acceptsFrom S ↔ y ∈ M.acceptsFrom (M.evalFrom S x) := by
+  simp [mem_acceptsFrom]
+
+theorem acceptsFrom_union {S1 S2 : Set σ} :
+    M.acceptsFrom (S1 ∪ S2) = M.acceptsFrom S1 + M.acceptsFrom S2 := by
+  rw [Language.add_def]; ext x
+  rw [Set.mem_union]; simp_rw [↑mem_acceptsFrom, evalFrom_union]
+  constructor
+  · rintro ⟨s, hs, h | h⟩
+    · left; tauto
+    · right; tauto
+  · rintro (⟨s, hs, h⟩ | ⟨s, hs, h⟩) <;> exists s <;> tauto
+
+theorem mem_acceptsFrom_biUnion {ι : Type*} (t : Set ι) (f : ι → Set σ) (x : List α) :
```

Comment:
Can you add the plain `iUnion` version too?

Reply:
(then the proof of this one is just to use iUnion twice)

Reply:
despite this snag I added `iUnion` lemas for `evalFrom` and `acceptsFrom`

Reply:
thanks for the catch! fixed